### DISCUSSION TITLE
Tracking JSI exceptions

### DISCUSSION THREAD

**Discussion by Minishlink** [invertase/react-native-firebase#5348] [Ideas]
Hello,

We're using reanimated 2, and we had quite a big bug when going in production with it. It raised lots of "libc.so" crashes with which we cannot do anything. After efforts, we successfully reproduced the problem on a device, and in the logs I see something like this, just before the SIGABRT :

```
l@dac0b50[MainActivity]  I  ViewPostIme pointer 0
               HermesVM  I  JSI rethrowing JS exception: unknown hex color: -30304
                         I  Error: unknown hex color: -30304
                         I      at sr (/var/folders/wp/4pf4qdmn217djs68lskdcc780000gn/T/code-push2021421-4109-6houg9.xawbo/CodePush/index.android.bundle:1933:8499)
                         I      at apply (native)
                         I      at y (/var/folders/wp/4pf4qdmn217djs68lskdcc780000gn/T/code-push2021421-4109-6houg9.xawbo/CodePush/index.android.bundle:1933:1454)
                         I      at n (/var/folders/wp/4pf4qdmn217djs68lskdcc780000gn/T/code-push2021421-4109-6houg9.xawbo/CodePush/index.android.bundle:1933:1722)
                         I      at u (/var/folders/wp/4pf4qdmn217djs68lskdcc780000gn/T/code-push2021421-4109-6houg9.xawbo/CodePush/index.android.bundle:1861:296)
                         I      at f (/var/folders/wp/4pf4qdmn217djs68lskdcc780000gn/T/code-push2021421-4109-6houg9.xawbo/CodePush/index.android.bundle:1861:1569)
                   libc  F  Fatal signal 6 (SIGABRT), code -6 (SI_TKILL) in tid 28166 (mqt_js), pid 27295
```

So I was wondering if we could catch the JS exceptions rethrown by JSI and send them to crashlytics? What do you think?

**Comment by mikehardy** [comment]
I may be misunderstanding, but I thought this was what crashlytics NDK reporting was for?
https://firebase.google.com/docs/crashlytics/ndk-reports
https://rnfirebase.io/crashlytics/usage#crashlytics-ndk

Not sure if it is *still* enabled by default as indicated above, it may actually require a slightly different configuration - it has not been a need of mine so I have not explored it much sorry, but that is supposed to be how these are reported to crashlytics?

Of course there is also play store console but that's no fun as it is effectively a separate crash database and looking at two is a terrible workflow - I'm not suggesting it as a solution, just as a temporary idea

I would love to hear if - now that you can reproduce this locally - you can get crashlytics NDK reporting going, and if there are any documentation updates needed it would be fantastic if we could collaborate to get them up to date + correct

**Reply by Minishlink** [reply]
Ok so there's definitely something missing in the NDK reporting in our app because on Google Play Store I see two ANR with number of occurences that match the spike of the SIGABRT that we see on Crashlytics. These are better but not yet optimal, I wouldn't be able to get the root cause from this (`com.facebook.jni.UnknownCppException` but also a simple `abort`)

What I was talking about though was more about the root cause of these native crashes. The actual code that crashes looks like this:

```js
import { runOnJS } from 'react-native-reanimated';

const myJSFunction = () => {
  throw new Error('hello');
};

runOnJS(myJSFunction)();
```

I have not yet dig a lot, but what happens in my understanding is that `myJSFunction` is run from the UI thread, so that's why it triggers a SIGABRT when it crashes instead of a regular JS crash. What I would like to do is to capture this JS exception (the `HermesVM` part of the log on my first post) before the crash so that we can see it on Crashlytics - be it as a separate exception or a Crashlytics log.

**Reply by mikehardy** [reply]
Hmm, ANR is different then crash though. I don't believe ANR will be detected by crashlytics unless you add something like https://github.com/SalomonBrys/ANR-WatchDog and respond to it's hooks by logging in to crashlytics.

If as it appears from your analysis this is running on UI thread and hangs (due to ??, but triggers ANR report in play console) then the user sees the ANR dialog and kills the app, then I'm not sure how to handle that.

It is definitely possible to get JS exceptions logged though, I just did a huge amount of work there - lines of code small but feature addition/configuration-ability change large - have you seen https://invertase.io/blog/react-native-firebase-crashlytics-configuration ?

Perhaps having crashlytics set to log JS crashes with no exception chaining, and temporarily enabled in debug while you reproduce will result in the crash you expect?

The ANR thing is perhaps related but actually separate to all of that

**Reply by Minishlink** [reply]
Sorry yes the ANR thing is not related, the reports I've seen are actually "crashes" in Play Console. I don't why I said ANR ^^"

I have indeed seen your blog post and disabled exception chaining before putting the new version in production :) My firebase.json looks like this

```json
{
  "react-native": {
    "crashlytics_debug_enabled": false,
    "crashlytics_javascript_exception_handler_chaining_enabled": false,
    "crashlytics_ndk_enabled": true
  }
}
```

**Reply by mikehardy** [reply]
Turn it on in debug then and reproduce, not sure why these aren't working ?
